src/lex.cpp

```diff
@@ -1337,6 +1339,75 @@ void lexer::parse_number() {
     QLJS_SLOW_ASSERT(
         !(number_begin[0] == u8'0' && this->is_digit(number_begin[1])));
   }
+  if (!has_decimal_point && !has_exponent && !is_bigint) {
+    std::string cleaned_string = "";
+    for (char8 c : string8_view(number_begin, input - number_begin)) {
+      if (c != '_') {
+        cleaned_string.push_back(static_cast<char>(c));
+      }
+    }
+    if (cleaned_string.size() > 15) {
+      bool error = false;
+      bump_vector<char8, monotonic_allocator> num_pretty(
+          "opening_tag_name_pretty", &this->allocator_);
+      if (cleaned_string.size() > 309) {
+        num_pretty += u8"inf";
+        string8_view num_pretty_view(num_pretty);
+        this->diag_reporter_->report(diag_number_literal_will_lose_precision{
+            .characters = source_code_span(number_begin, input),
+            .rounded_val = num_pretty_view,
+        });
+        error = true;
+      }
+      if (!error) {
+        long double long_num = stold(cleaned_string);
+        double num = long_num;
+        std::array<char, 310> result_string;
+        int rc = std::snprintf(result_string.data(), result_string.size(),
+                               "%.0f", num);
+        if (rc < 0 || rc >= result_string.size()) {
+          // I think the inside of this if block is unreachable.
```

Comment:
If it's really unreachable, call `QLJS_UNREACHABLE`, or write `QLJS_ALWAYS_ASSERT(rc < 0 || rc >= result_string.size())`.